core/modules/wsched/ChunkTaskList.cc

```diff
@@ -0,0 +1,393 @@
+    // will not examine any further chunks upon seeing those results.
+    auto task = _activeTasks.top();
+    if (!task->hasMemHandle()) {
+        memman::TableInfo::LockType lckOptTbl = memman::TableInfo::LockType::MUSTLOCK;
+        memman::TableInfo::LockType lckOptIdx = memman::TableInfo::LockType::NOLOCK;
+        if (useFlexibleLock) lckOptTbl = memman::TableInfo::LockType::FLEXIBLE;
+        auto scanInfo = task-> getScanInfo();
+        auto chunkId = task->getChunkId();
+        if (chunkId != _chunkId) {
+            // This would slow things down badly, but the system would survive.
+            LOGS(_log, LOG_LVL_ERROR, "ChunkTasks " << _chunkId << " got task for chunk " << chunkId
+                    << " " << task->getIdStr());
+        }
+        std::vector<memman::TableInfo> tblVect;
+        for (auto const& tbl : scanInfo.infoTables) {
+            memman::TableInfo ti(tbl.db + "/" + tbl.table, lckOptTbl, lckOptIdx);
+            tblVect.push_back(ti);
+        }
+        // If tblVect is empty, we should get the empty handle
+        memman::MemMan::Handle handle = _memMan->lock(tblVect, chunkId);
+        if (handle == 0) {
+            switch (errno) {
+            case ENOMEM:
+                logMemManRes(true, "ENOMEM", tblVect);
+                return ChunkTasks::ReadyState::NO_RESOURCES;
+            case ENOENT:
+                LOGS(_log, LOG_LVL_ERROR, "_memMgr->lock errno=ENOENT chunk not found " << task->getIdStr());
+                // Not sure if this is the best course of action, but it should just need one
+                // logic path. The query should fail from the missing tables
+                // and the czar must be able to handle that with appropriate retries.
+                handle = memman::MemMan::HandleType::ISEMPTY;
+                break;
+            default:
+                LOGS(_log, LOG_LVL_ERROR, "_memMgr->lock file system error " << task->getIdStr());
+                // Any error reading the file system is probably fatal for the worker.
+                throw std::bad_exception();
+                return ChunkTasks::ReadyState::NO_RESOURCES;
+            }
+        }
+        task->setMemHandle(handle);
+        logMemManRes(false, task->getIdStr() + " got handle", tblVect);
+    }
+
+    // There is a Task to run at this point, pull it off the heap to avoid confusion.
+    auto popped = _activeTasks.pop();
+    _readyTask = task;
+    if (popped != task) {
+        // This would be deadly.
+        throw Bug(std::string("ChunkTasks::ready popped and task don't match! ") +
+                "task=" + task->getIdStr() + " popped=" + popped->getIdStr());
+    }
+    return ChunkTasks::ReadyState::READY;
+}
+
+
+int ChunkTaskList::getActiveChunkId() {
+    if (_activeChunk == _chunkMap.end()) {
+        return -1;
+    }
+    return _activeChunk->second->getChunkId();
+}
+
+
+/// @return old value of _resourceStarved.
```

Comment:
If this is used only for logging purpose, maybe you could add it in documentation?